db/database_test.go

```diff
@@ -1907,3 +1908,179 @@ func TestSyncFnMutateBody(t *testing.T) {
+
+	body = Body{"name": "Harry", "age": 40}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 2-a")
+
+	body = Body{"name": "Amelia", "age": 20}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"3-a", "2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 3-a")
+
+	body = Body{"name": "Charlie", "age": 10}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"4-a", "3-a", "2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 4-a")
+
+	body = Body{"name": "Noah", "age": 40}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-b", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 2-b")
+
+	var wg sync.WaitGroup
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			body := Body{"name": "Emily", "age": 20}
+			_, _, err := db.PutExistingRevWithBody("doc1", body, []string{"3-b", "2-b", "1-a"}, false)
+			assert.NoError(t, err, "Adding revision 3-b")
+		}()
+	}
+	wg.Wait()
+
+	doc, err := db.GetDocument("doc1", DocUnmarshalAll)
+	assert.NoError(t, err, "Retrieve doc after adding 3-b")
+	assert.Equal(t, "4-a", doc.CurrentRev)
+}
+
+// Multiple clients are attempting to push the same tombstone of the winning revision for a branched document
+// * First writer should be successful, subsequent writers should fail on CAS, then identify rev already exists
+func TestConcurrentPushSameTombstoneWinningRevision(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeyAll)()
+	db, testBucket := setupTestDB(t)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	body := Body{"name": "Olivia", "age": 80}
+	_, _, err := db.PutExistingRevWithBody("doc1", body, []string{"1-a"}, false)
+	assert.NoError(t, err, "Adding revision 1-a")
+
+	body = Body{"name": "Harry", "age": 40}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 2-a")
+
+	body = Body{"name": "Amelia", "age": 20}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"3-a", "2-a", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 3-a")
+
+	body = Body{"name": "Noah", "age": 40}
+	_, _, err = db.PutExistingRevWithBody("doc1", body, []string{"2-b", "1-a"}, false)
+	assert.NoError(t, err, "Adding revision 2-b")
+
+	doc, err := db.GetDocument("doc1", DocUnmarshalAll)
+	assert.NoError(t, err, "Retrieve doc before tombstone")
+	assert.Equal(t, "3-a", doc.CurrentRev)
+
+	success, failure := int64(0), int64(0)
+	var wg sync.WaitGroup
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			body := Body{"name": "Charlie", "age": 10, BodyDeleted: true}
+			_, _, err := db.PutExistingRevWithBody("doc1", body, []string{"4-a", "3-a", "2-a", "1-a"}, false)
+			assert.NoError(t, err, "Adding revision 4-a (tombstone)")
+
+			if err != nil {
+				assert.Equal(t, "409 Document exists", err.Error())
```

Comment:
Since we don't expect the failure case in this test, does this assert make sense?